File: claude-code-CLI/src/orchestration/executionState.ts
/**
 * Execution State API client for Claude Code CLI.
 * Connects to the local proxy server to query and manage persistent orchestration state.
 */

import { getParentSessionId, getSessionId } from '../bootstrap/state.js'
import { getAuthHeaders } from '../utils/http.js'

export type ExecutionPhase =
	| 'idle'
	| 'backend_planning'
	| 'backend_execution'
	| 'backend_validation'
	| 'frontend_planning'
	| 'frontend_execution'
	| 'frontend_validation'
	| 'integration_check'

export interface PlanStep {
	step_id: string
	description: string
	status: 'pending' | 'in_progress' | 'completed' | 'skipped'
	completed_at: string | null
}

export interface CheckpointState {
	name: string
	description: string
	phase: ExecutionPhase
	created_at: string
}

export interface ExecutionState {
	session_id: string
	parent_session_id?: string | null
	active_model: string
	current_checkpoint: CheckpointState | null
	implementation_phase: ExecutionPhase
	approved_plan: string | null
	completed_steps: PlanStep[]
	remaining_steps: PlanStep[]
	locked_rules: string[]
	active_files: string[]
	validation_findings: string[]
	last_updated: string
	version: number
}

export interface ExecutionStateResponse extends Partial<ExecutionState> {
	exists: boolean
}

function getExecutionStateBaseUrl(): string | null {
	if (!isExecutionStateOrchestrationEnabled()) return null
	const baseUrl = process.env.ANTHROPIC_BASE_URL
	if (!baseUrl) return null
	return baseUrl.replace(/\/+$/, '').replace(/\/v1$/, '')
}

export function isExecutionStateOrchestrationEnabled(): boolean {
	const value = process.env.CLAUDE_CODE_ENABLE_EXECUTION_STATE_ORCHESTRATION
	if (!value) return false
	return ['1', 'true', 'yes', 'on'].includes(value.trim().toLowerCase())
}

function getExecutionStateHeaders(): Record<string, string> | null {
	const auth = getAuthHeaders()
	if (auth.error) return null

	const sessionId = getSessionId()
	const parentSessionId = getParentSessionId()

	return {
		...auth.headers,
		'x-session-id': sessionId,
		...(parentSessionId ? { 'x-parent-session-id': parentSessionId } : {}),
	}
}

/**
 * Get the full execution state for the current session.
 */
export async function fetchExecutionState(): Promise<ExecutionStateResponse | null> {
	try {
		const proxyUrl = getExecutionStateBaseUrl()
		if (!proxyUrl) return null
		const sessionId = getSessionId()
		const headers = getExecutionStateHeaders()
		if (!headers) return null

		const response = await fetch(`${proxyUrl}/v1/execution_state/${sessionId}`, {
			method: 'GET',
			headers: {
				...headers,
				Accept: 'application/json',
			},
		})

		if (!response.ok) {
			if (response.status === 404) return null
			throw new Error(`Failed to fetch execution state: ${response.statusText}`)
		}

		return (await response.json()) as ExecutionStateResponse
	} catch {
		// Graceful fallback if proxy is unavailable or endpoint missing
		return null
	}
}

/**
 * Partially update the execution state.
 */
export async function updateExecutionState(
	update: Partial<ExecutionState>,
): Promise<ExecutionState | null> {
	try {
		const proxyUrl = getExecutionStateBaseUrl()
		if (!proxyUrl) return null
		const sessionId = getSessionId()
		const headers = getExecutionStateHeaders()
		if (!headers) return null

		const response = await fetch(`${proxyUrl}/v1/execution_state/${sessionId}`, {
			method: 'PUT',
			headers: {
				...headers,
				'Content-Type': 'application/json',
				Accept: 'application/json',
			},
			body: JSON.stringify(update),
		})

		if (!response.ok) {
			throw new Error(`Failed to update execution state: ${response.statusText}`)
		}

		return (await response.json()) as ExecutionState
	} catch {
		return null
	}
}

/**
 * Parse and persist an approved execution plan.
 */
export async function applyPlan(planText: string): Promise<ExecutionState | null> {
	try {
		const proxyUrl = getExecutionStateBaseUrl()
		if (!proxyUrl) return null
		const sessionId = getSessionId()
		const headers = getExecutionStateHeaders()
		if (!headers) return null
		const parentSessionId = getParentSessionId()

		const response = await fetch(`${proxyUrl}/v1/execution_state/${sessionId}/apply_plan`, {
			method: 'POST',
			headers: {
				...headers,
				'Content-Type': 'application/json',
				Accept: 'application/json',
			},
			body: JSON.stringify({
				plan_text: planText,
				...(parentSessionId ? { parent_session_id: parentSessionId } : {}),
			}),
		})

		if (!response.ok) {
			throw new Error(`Failed to apply plan: ${response.statusText}`)
		}

		return (await response.json()) as ExecutionState
	} catch {
		return null
	}
}

/**
 * Mark a specific step as completed in the plan.
 */
export async function completePlanStep(stepId: string): Promise<ExecutionState | null> {
	try {
		const proxyUrl = getExecutionStateBaseUrl()
		if (!proxyUrl) return null
		const sessionId = getSessionId()
		const headers = getExecutionStateHeaders()
		if (!headers) return null

		const response = await fetch(`${proxyUrl}/v1/execution_state/${sessionId}/complete_step`, {
			method: 'POST',
			headers: {
				...headers,
				'Content-Type': 'application/json',
				Accept: 'application/json',
			},
			body: JSON.stringify({ step_id: stepId }),
		})

		if (!response.ok) {
			throw new Error(`Failed to complete step: ${response.statusText}`)
		}

		return (await response.json()) as ExecutionState
	} catch {
		return null
	}
}

/**
 * Set a checkpoint in the current session.
 */
export async function setCheckpoint(
	name: string,
	description = '',
	phase?: ExecutionPhase,
): Promise<ExecutionState | null> {
	try {
		const proxyUrl = getExecutionStateBaseUrl()
		if (!proxyUrl) return null
		const sessionId = getSessionId()
		const headers = getExecutionStateHeaders()
		if (!headers) return null

		const response = await fetch(`${proxyUrl}/v1/execution_state/${sessionId}/set_checkpoint`, {
			method: 'POST',
			headers: {
				...headers,
				'Content-Type': 'application/json',
				Accept: 'application/json',
			},
			body: JSON.stringify({ name, description, phase }),
		})

		if (!response.ok) {
			throw new Error(`Failed to set checkpoint: ${response.statusText}`)
		}

		return (await response.json()) as ExecutionState
	} catch {
		return null
	}
}

/**
 * Restore a checkpoint in the current session.
 */
export async function restoreCheckpoint(): Promise<ExecutionState | null> {
	try {
		const proxyUrl = getExecutionStateBaseUrl()
		if (!proxyUrl) return null
		const sessionId = getSessionId()
		const headers = getExecutionStateHeaders()
		if (!headers) return null

		const response = await fetch(`${proxyUrl}/v1/execution_state/${sessionId}/restore_checkpoint`, {
			method: 'POST',
			headers: {
				...headers,
				Accept: 'application/json',
			},
		})

		if (!response.ok) {
			throw new Error(`Failed to restore checkpoint: ${response.statusText}`)
		}

		return (await response.json()) as ExecutionState
	} catch {
		return null
	}
}

File: claude-code-proxy-server/api/models/execution_state.py
"""Strongly typed models for persistent execution-state orchestration.

These models represent the complete orchestration state that survives
model switches, session resumes, and server restarts.  They are the
single source of truth consumed by the state store, the context
injector, and the CLI sync layer.
"""

from __future__ import annotations

from datetime import datetime, timezone
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field


class ExecutionPhase(StrEnum):
    """Lifecycle phases for implementation execution."""

    idle = "idle"
    backend_planning = "backend_planning"
    backend_execution = "backend_execution"
    backend_validation = "backend_validation"
    frontend_planning = "frontend_planning"
    frontend_execution = "frontend_execution"
    frontend_validation = "frontend_validation"
    integration_check = "integration_check"


class PlanStep(BaseModel):
    """A single step in an implementation plan."""

    step_id: str
    description: str
    status: Literal["pending", "in_progress", "completed", "skipped"] = "pending"
    completed_at: datetime | None = None

    def mark_completed(self) -> PlanStep:
        """Return a copy marked as completed with the current timestamp."""
        return self.model_copy(
            update={
                "status": "completed",
                "completed_at": datetime.now(timezone.utc),
            }
        )

    def mark_in_progress(self) -> PlanStep:
        """Return a copy marked as in-progress."""
        return self.model_copy(update={"status": "in_progress"})


class CheckpointState(BaseModel):
    """A named checkpoint within an execution session."""

    name: str
    description: str = ""
    phase: ExecutionPhase = ExecutionPhase.idle
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class ExecutionState(BaseModel):
    """Complete persistent orchestration state for a session.

    This is persisted as JSON and injected into provider requests so
    that any model — regardless of provider — can seamlessly continue
    execution from the last known state.
    """

    session_id: str
    active_model: str = ""
    current_checkpoint: CheckpointState | None = None
    implementation_phase: ExecutionPhase = ExecutionPhase.idle
    approved_plan: str | None = None
    completed_steps: list[PlanStep] = Field(default_factory=list)
    remaining_steps: list[PlanStep] = Field(default_factory=list)
    locked_rules: list[str] = Field(default_factory=list)
    active_files: list[str] = Field(default_factory=list)
    architecture_sources: list[str] = Field(default_factory=list)
    validation_findings: list[str] = Field(default_factory=list)
    last_updated: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    version: int = 1

    def touch(self) -> None:
        """Update the ``last_updated`` timestamp in place."""
        self.last_updated = datetime.now(timezone.utc)

    def progress_summary(self) -> tuple[int, int]:
        """Return ``(completed_count, total_count)``."""
        completed = sum(1 for s in self.completed_steps if s.status == "completed")
        total = len(self.completed_steps) + len(self.remaining_steps)
        return completed, total

    def has_active_plan(self) -> bool:
        """Return whether an approved plan with steps exists."""
        return self.approved_plan is not None and (
            bool(self.completed_steps) or bool(self.remaining_steps)
        )


class ExecutionStateUpdate(BaseModel):
    """Partial update payload accepted by the REST API.

    Only non-None fields are applied to the existing state.
    """

    active_model: str | None = None
    current_checkpoint: CheckpointState | None = None
    implementation_phase: ExecutionPhase | None = None
    approved_plan: str | None = None
    completed_steps: list[PlanStep] | None = None
    remaining_steps: list[PlanStep] | None = None
    locked_rules: list[str] | None = None
    active_files: list[str] | None = None
    architecture_sources: list[str] | None = None
    validation_findings: list[str] | None = None
